syntax = "proto3";

message TestRootMsg {
    Param param = 1;
    EData evalData = 2;
    APISeq apiSequence = 3;
}

message Param {
    enum SecretKeyDist {
        GAUSSIAN        = 0;
        UNIFORM_TERNARY = 1;
        SPARSE_TERNARY  = 2;
    }
    optional int32 num32 = 1; 
    int64 num64 = 2;
    optional SecretKeyDist secretKeyDist = 3;
}

message EData {
    message OneDataList {
        repeated double dataList = 1;
    }
    uint32 len = 1;   // length of each data list
    repeated OneDataList allDataLists = 2;
}

message APISeq {
    message OneAPI {
        message AddTwoList {
            uint32 src1 = 1;
            uint32 src2 = 2;
        }
        message AddManyList {
            repeated uint32 srcs = 1;
        }
        message MulTwoList {
            uint32 src1 = 1;
            uint32 src2 = 2;
        }
        message MulManyList {
            repeated uint32 srcs = 1;
        }
        message ShiftOneList {
            uint32 src = 1;
            uint32 index = 2;
        }    
        oneof api{
            AddTwoList addTwoList = 1;
            AddManyList addManyList = 2;
            MulTwoList mulTwoList = 3;
            MulManyList mulManyList = 4;
            ShiftOneList shiftOneList = 5;
        };
        uint32 dst = 6;
    }
    OneAPI testone = 2;
    repeated OneAPI apiList = 1;
}